### PR TITLE
Implement undo for add and delete and edit (list of commands support)

### PR DESCRIPTION
**Pull request by turtle96**
to view list of history use "list-undo"
then use "undo INDEX"

### CODE REVIEW

**Comment by turtle96** [comment]
@CS2103AUG2016-W15-C3/documentation 
@CS2103AUG2016-W15-C3/code-quality 
@CS2103AUG2016-W15-C3/testing 

Please review

File: src/main/java/seedu/taskell/logic/commands/DeleteCommand.java [hunk 1]
```diff
@@ -40,6 +41,7 @@ public CommandResult execute() {
 
         try {
             model.deleteTask(taskToDelete);
+            UndoCommand.addTaskToCommandHistory((Task) taskToDelete);
         } catch (TaskNotFoundException pnfe) {
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/DeleteCommand.java]: Why is it pnfe? Should it be tnfe, which refers to TaskNotFoundException

File: src/main/java/seedu/taskell/logic/commands/UndoCommand.java [hunk 2]
```diff
@@ -4,42 +4,66 @@
...
+            + "Example: " + COMMAND_WORD + " 3";
     
-    public static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
-    public static final String MESSAGE_ADD_TASK_SUCCESS = "Task added back: %1$s";
-    public static final String MESSAGE_DUPLICATE_TASK = "This task already exists in the task manager";
-    public static final String MESSAGE_NO_TASK_TO_UNDO = "No add or delete commands available to undo.";
-
-    private static ArrayList<String> commandHistory;
+    private static final String MESSAGE_DELETE_TASK_SUCCESS = "Deleted Task: %1$s";
+    private static final String MESSAGE_ADD_TASK_SUCCESS = "Task added back: %1$s";
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/UndoCommand.java:28]: Can use the constant that has already been declared in AddCommand.java

File: src/main/java/seedu/taskell/logic/commands/UndoCommand.java [hunk 3]
```diff
@@ -4,42 +4,66 @@
...
     }
     
     @Override
     public CommandResult execute() {
-        if (mostRecentCommand==null) {
-            return new CommandResult(String.format(MESSAGE_NO_TASK_TO_UNDO));
+        
+        if (commandHistoryList.isEmpty()) {
+            return new CommandResult(String.format(MESSAGE_COMMAND_HISTORY_EMPTY));
         }
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/UndoCommand.java]: else if should process after '{'

image

File: src/main/java/seedu/taskell/logic/commands/UndoCommand.java [hunk 4]
```diff
@@ -48,45 +72,64 @@ public CommandResult execute() {
...
             return new CommandResult(MESSAGE_DUPLICATE_TASK);
         }
     }
 
     private CommandResult undoAdd() {
         try {
-            model.deleteTask(mostRecentAddedTask);
+            model.deleteTask(commandHistory.getTask());
+            deleteCommandHistory();
         } catch (TaskNotFoundException e) {
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/UndoCommand.java:167]: Inconsistent naming of exception
image
.

File: src/main/java/seedu/taskell/logic/commands/UndoCommand.java [hunk 5]
```diff
@@ -48,45 +72,64 @@ public CommandResult execute() {
...
-        mostRecentCommand = commandWord;
+
+    private void deleteCommandHistory() {
+        commandHistoryList.remove(commandHistory);
     }
     
-    public static void updateMostRecentAddedTask(Task task) {
-        mostRecentAddedTask = task;
+    public static void initializeCommandHistory() {
+        if (commandHistoryList==null) {
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/UndoCommand.java]: Whitespace should be introduced.
`if (commandHistoryList == null) {`
image

**Comment by marcusngwenjian** [comment]
Code Quality reviewed

Refer to #98 for bugs found

**Comment by marcusngwenjian** [comment]
Refer to issue #98 Figure 3 for new found bug.

**Comment by turtle96** [comment]
This fixes #98 and closes #98 
@ZuoZhuolin  
please check

File: src/main/java/seedu/taskell/logic/commands/UndoCommand.java [hunk 6]
```diff
@@ -1,92 +1,243 @@
...
     
     @Override
     public CommandResult execute() {
-        if (mostRecentCommand==null) {
-            return new CommandResult(String.format(MESSAGE_NO_TASK_TO_UNDO));
+        
+        if (commandHistoryList.isEmpty()) {
+            return new CommandResult(String.format(MESSAGE_COMMAND_HISTORY_EMPTY));
+        }
+        else if (index > commandHistoryList.size()) {
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/UndoCommand.java]: else if should comes after '{'
image

File: src/main/java/seedu/taskell/logic/commands/UndoCommand.java [hunk 7]
```diff
@@ -1,92 +1,243 @@
...
+        commandHistory.setToRedoToTrue();
+        commandHistoryList.add(commandHistory);
     }
     
-    public static void updateMostRecentAddedTask(Task task) {
-        mostRecentAddedTask = task;
+    /******** static methods *********/
+    
+    public static void initializeCommandHistory() {
+        if (commandHistoryList==null) {
```
Comment (marcusngwenjian) [inline src/main/java/seedu/taskell/logic/commands/UndoCommand.java]: Whitespace required
image

image

**Comment by marcusngwenjian** [comment]
Code quality reviewed.

Please refer to issue #109 for new bug found

**Comment by Mamtha3005** [comment]
Documentation Reviewed.

**Comment by turtle96** [comment]
closes #53

**Comment by ZuoZhuolin** [comment]
After clear command, there are still some items in undo list, but the items can not work.

**Comment by turtle96** [comment]
closes #108 
closes #53
fix bug #98

**Comment by turtle96** [comment]
closes #53

**Pull request merged**